src/GraphQL.Federation/Instrumentation/FederatedTraceBuilder.cs

```diff
@@ -0,0 +1,238 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+using GraphQL.Execution;
+using GraphQL.Instrumentation;
+using GraphQL.Validation;
+using Mdg.Engine.Proto;
+using static Mdg.Engine.Proto.Trace.Types;
+
+namespace GraphQL.Federation.Instrumentation
+{
+    /// <summary>
+    /// Builds a proto tree structure with the tracing data based on the parent-child relationship
+    /// among fields.
+    /// </summary>
+    public class FederatedTraceBuilder
+    {
+        private readonly PerfRecord[] _records;
+        private readonly ExecutionErrors _errors;
+        private readonly DateTime _start;
+        /// <summary>
+        /// Instantiates trace builder
+        /// </summary>
+        /// <param name="records">tracing data captured at parse/validate/execute steps</param>
+        /// <param name="errors">errors if occurred during parse and validate step</param>
+        /// <param name="start">process start time in UTC</param>
+        public FederatedTraceBuilder(PerfRecord []records, ExecutionErrors errors, DateTime start)
+        {
+            _records = records;
+            _errors = errors;
+            _start = start;
+        }
```

Comment:
```suggestion
        }
        
```

NewLine after each method.